helix-view/src/editor.rs

```diff
@@ -71,6 +72,105 @@ where
     )
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case", default, deny_unknown_fields)]
+pub struct GutterConfig {
+    /// Gutter Layout
+    pub layout: Vec<GutterType>,
+    /// Options specific to the "line-numbers" gutter
+    pub line_numbers: GutterLineNumbersConfig,
+}
+
+impl Default for GutterConfig {
+    fn default() -> Self {
+        Self {
+            layout: vec![
+                GutterType::Diagnostics,
+                GutterType::Spacer,
+                GutterType::LineNumbers,
+                GutterType::Spacer,
+                GutterType::Diff,
+            ],
+            line_numbers: GutterLineNumbersConfig::default(),
+        }
+    }
+}
+
+impl From<Vec<GutterType>> for GutterConfig {
+    fn from(x: Vec<GutterType>) -> Self {
+        GutterConfig {
+            layout: x,
+            ..Default::default()
+        }
+    }
+}
+
+fn deserialize_gutter_seq_or_struct<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+where
+    T: Deserialize<'de> + From<Vec<GutterType>>,
+    D: Deserializer<'de>,
+{
+    // This is a Visitor that forwards string types to T's `FromStr` impl and
+    // forwards map types to T's `Deserialize` impl. The `PhantomData` is to
+    // keep the compiler from complaining about T being an unused generic type
+    // parameter. We need T in order to know the Value type for the Visitor
+    // impl.
+    struct GutterSeqOrStruct<T>(PhantomData<fn() -> T>);
```

Comment:
You can probably use `*const T`, which is also covariant in `T`.

Reply:
Admittedly I'm just getting exposed to this now - but I appreciate the terminology which prompted a nice reading on variance in the nomicon. 

In this case, this type signature (and the accompanying comment) is lifted straight from the example provided in the [serde docs](https://serde.rs/string-or-struct.html). If there are merits to one or the other, it's not obvious to me.

`fn() -> *const T` isn't covered in [nomicon's page on PhantomData patterns](https://doc.rust-lang.org/nomicon/phantom-data.html#table-of-phantomdata-patterns). In contrast to `PhantomData<T>` vs `PhantomData<*const T>` recommended [here](https://doc.rust-lang.org/core/marker/struct.PhantomData.html#ownership-and-the-drop-check), I'm not seeing similar discussion regarding `fn() -> T` vs `fn() -> *const T`. 

I'm happy to make the change, but honestly don't know enough about the implications of each. If they provide the same variance, does using `fn() -> *const T` imbue some other benefit over `fn() -> T`?

Reply:
The implication is the same, since both `*const T` and `fn() -> T` are covariant in `T` (`fn(T)` is contravariant in `T` though). Covariance and contravariance can be a bit overwhelming are basically used to know how subtyping can be done. IIRC rustc uses it to implement coercion rules (and maybe drop but I’m not sure?).

Either way I think it should be okay, I was just mentioning it.

Reply:
Okay cool - then I'd default to whatever is recommended in the serde docs. Let's leave it as is and if there's ever a strong rationale to switch that can always come down the road in another PR.